scripts/js/upgrade/index.js

```diff
@@ -0,0 +1,330 @@
+const { ApiPromise, WsProvider, Keyring } = require('@polkadot/api');
+const { blake2AsHex } = require('@polkadot/util-crypto');
+const fs = require('fs')
+const util = require('util');
+const exec = util.promisify(require('child_process').exec);
+
+const run = async () => {
+  let exitCode = 0;
+  try {
+    console.log("Parsing Args ...")
+    // 0 & 1 are command context
+    const endpoint = "ws://0.0.0.0:9946";
+    const endpointRelay = "ws://0.0.0.0:9944";
+    const seeds = ["//Alice", "//Bob", "//Charlie"];
+    const wasmFile = process.argv[2];
+    const targetDockerTag = process.argv[3];
+
+    const wsProvider = new WsProvider(endpoint);
+    const wsProviderRelay = new WsProvider(endpointRelay);
+
+    console.log("Starting Relay Chain and waiting until is up")
+    await execCommand('cd ../../../ && ./scripts/init.sh start-relay-chain')
+
+    const apiRelay = await ApiPromise.create({
+      provider: wsProviderRelay,
+    });
+
+    console.log("Waiting until relay chain is producing blocks")
+    await waitUntilEventFound(apiRelay, "ExtrinsicSuccess")
+
+    console.log("Starting Centrifuge Chain and waiting until is up")
+    process.env.CC_DOCKER_TAG = targetDockerTag
+    process.env.PARA_CHAIN_SPEC = "centrifuge-local" // Make it configurable
+    await execCommand('cd ../../../ && ./scripts/init.sh start-parachain-docker')
+
+    const api = await ApiPromise.create({
+      provider: wsProvider,
+    });
+
+    console.log("Centrifuge Chain started, onboarding parachain now")
+    process.env.DOCKER_ONBOARD = true
+    process.env.PARA_DOCKER_IMAGE_TAG = targetDockerTag
+    await execCommand('cd ../../../ && ./scripts/init.sh onboard-parachain')
+
+    console.log("Waiting until Centrifuge Chain is producing blocks")
+    await waitUntilEventFound(api, "ExtrinsicSuccess")
+
+    // Wait one extra session
+    await waitUntilEventFound(api, "EmptyTerm")
+    // await waitUntilEventFound(api, "NewSession")
+
+    const keyring = new Keyring({ type: "sr25519" });
+    const alice = keyring.addFromUri(seeds[0]);
+    const bob = keyring.addFromUri(seeds[1]);
+    const charlie = keyring.addFromUri(seeds[2]);
+
+    const wasm = fs.readFileSync(wasmFile)
+    const wasmHash = blake2AsHex(wasm)
+    console.log("Applying WASM Blake2 Hash:", wasmHash)
+
+    let nonce = Number((await api.query.system.account(alice.address)).nonce);
+    const preimageNoted = await notePreimageAuth(api, alice, wasmHash, nonce);
+
+    console.log("Continuing with council proposal using", preimageNoted)
+    nonce = Number((await api.query.system.account(alice.address)).nonce);
```

Comment:
NIT but shall we extract this as a function `get_nonce(api, address)` just to reduce some duplication and noise?